feat(volo-http): support merging routers

yukiiiteru opened the pull request:
## Motivation

When developing a large project, users can create a `Router` in each mod and merge them in main mod. Merging routers is an important feature for this situation.

Additionally, this PR updates the http example with split routers, and more types support the `FromRequest` and `IntoResponse` traits.

yukiiiteru committed:
feat(volo-http): support router merge

Signed-off-by: Yu Li <liyu.yukiteru@bytedance.com>

yukiiiteru committed:
feat(volo-http): support `IntoResponse` for `Json`

Signed-off-by: Yu Li <liyu.yukiteru@bytedance.com>

yukiiiteru committed:
chore(volo-http): update examples for http

Signed-off-by: Yu Li <liyu.yukiteru@bytedance.com>

PureWhiteWu commented on volo-http/src/extract.rs:
```diff
@@ -118,6 +118,21 @@ impl<S: Sync> FromRequest<S> for Incoming {
+
+    async fn from(_cx: &HttpContext, body: Incoming, _state: &S) -> Result<Self, Self::Rejection> {
+        // `Infallible` for `RespBody`
+        let body = body
```
Maybe we can check if we have `content-length` in the header, and pre-allocate something like `BytesMut` and read the body to it directly?
`collect` is backed by a `buflist`, and we need to call `to_bytes` which may lead to an extra mem copy.

wfly1998 replied:
Sounds good, I'll update it later.

PureWhiteWu reviewed (commented):
Also, `FastStr` is preferred instead of `String`.

wfly1998 commented:
> Also, `FastStr` is preferred instead of `String`.

The `String` is only used for merging routers currently, so performance does not matter here.

PureWhiteWu commented on volo-http/src/extract.rs:
```diff
@@ -118,6 +148,68 @@ impl<S: Sync> FromRequest<S> for Incoming {
+    }
+}
+
+impl<S: Sync> FromRequest<S> for UTF8String {
```
这里我的建议还是做成 String，不要分两个类型，然后使用 feature 来控制会比较好（默认校验）。
原因是：
1. String 本身就必须是 UTF-8 Valid 的，所以这里加个`UTF8String`本身就很奇怪；
2. `UncheckedString`虽然命名上为 Unchecked，但是实际在代码中并没有出现 unsafe，其实也还是违背了 Rust 的安全准则。
另外，也可以为 FastStr 实现一下。

yukiiiteru committed:
feat(volo-http): support `FromRequest` for more types

This commit supports `FromRequest` for `Vec<u8>`, `Bytes`, `String`
and `FastStr`.

Signed-off-by: Yu Li <liyu.yukiteru@bytedance.com>

PureWhiteWu commented on volo-http/src/extract.rs:
```diff
@@ -118,51 +120,119 @@ impl<S: Sync> FromRequest<S> for Incoming {
+            while let Some(next) = body.frame().await {
+                let frame = next.map_err(|_| RejectionError::BodyCollectionError)?;
+                let bytes = frame
+                    .into_data()
```
这里如果是 trailers 也不应该报错吧？以及我们是不是也要把 trailers 存下来？

wfly1998 replied:
经过调研发现，为了支持 HTTP/2，直接使用 `collect` 就是最优解，不需要做额外处理，所以删掉了这部分。

PureWhiteWu commented on volo-http/src/extract.rs:
```diff
@@ -118,51 +120,119 @@ impl<S: Sync> FromRequest<S> for Incoming {
+                let bytes = frame
+                    .into_data()
+                    .map_err(|_| RejectionError::BodyCollectionError)?;
+                vec.put(bytes);
```
这里是不是也要判断一下长度是否符合 content length，不符合就报错

wfly1998 replied:
现在添加了校验，但我认为如果拿到数据的长度与 content length 不符，也不应该报错，目前只会打出一条 warn。

PureWhiteWu commented on volo-http/src/extract.rs:
```diff
@@ -118,51 +120,119 @@ impl<S: Sync> FromRequest<S> for Incoming {
+    type Rejection = RejectionError;
+
+    async fn from(cx: &HttpContext, body: Incoming, state: &S) -> Result<Self, Self::Rejection> {
+        <Vec<u8> as FromRequest<S>>::from(cx, body, state)
```
这里感觉可以再优化一下，先判断一下有没有 content length header，有的话走我们自己的 Vec u8，没有的话直接走 body.collect 那个，要不然就会先从 bytes 转成了 vecu8 再转回去，可能会多一次 malloc 和 copy

PureWhiteWu commented on volo-http/src/extract.rs:
```diff
@@ -118,51 +120,119 @@ impl<S: Sync> FromRequest<S> for Incoming {
+    type Rejection = RejectionError;
+
+    async fn from(cx: &HttpContext, body: Incoming, state: &S) -> Result<Self, Self::Rejection> {
+        let vec = <Vec<u8> as FromRequest<S>>::from(cx, body, state).await?;
```
FastStr 可以从上面优化后的 Bytes 去转，因为 FastStr 不要求 mutable；String 不这么干是因为 String 只能从 Vec<u8> 无损转换得到

PureWhiteWu commented on volo-http/src/extract.rs:
```diff
@@ -130,28 +145,39 @@ where
-            }
-        }
+        let body = body
+            .collect()
```
这里同上，Json 的 Deserialize 可以从 Bytes 转性能会比较好

yukiiiteru committed:
feat(volo-http): add `MaybeInvalid` for string types

With `MaybeInvalid` extractor, a function can receive a string type
as argument and get it without UTF-8 checking by `assume_valid`.

Signed-off-by: Yu Li <liyu.yukiteru@bytedance.com>
